src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs

```diff
@@ -39,17 +49,54 @@ public override bool TryRead(Type valueType, ref Utf8JsonReader reader, out TVal
                 return false;
             }
 
-            if (s_isUint64)
+            if (s_isUInt64 && reader.TryGetUInt64(out ulong uint64))
             {
-                if (reader.TryGetUInt64(out ulong ulongValue))
-                {
-                    value = (TValue)Enum.ToObject(valueType, ulongValue);
-                    return true;
-                }
+                value = (TValue)Enum.ToObject(valueType, uint64);
+                return true;
+            }
+
+            if (s_isInt64 && reader.TryGetInt64(out long int64))
+            {
+                value = (TValue)Enum.ToObject(valueType, int64);
+                return true;
+            }
+
+            if (s_isUInt32 && reader.TryGetUInt32(out uint uint32))
+            {
+                value = (TValue)Enum.ToObject(valueType, uint32);
+                return true;
+            }
+
+            if (s_isInt32 && reader.TryGetInt32(out int int32))
+            {
+                value = (TValue)Enum.ToObject(valueType, int32);
+                return true;
+            }
+
+            // When utf8reader/writer will support all primitive types we should remove custom bound checks
+            // https://github.com/dotnet/corefx/issues/36125
+
+            if (s_isUInt16 && reader.TryGetUInt32(out uint uint16) && uint16 >= ushort.MinValue && uint16 <= ushort.MaxValue)
+            {
+                value = (TValue)Enum.ToObject(valueType, uint16);
+                return true;
+            }
+
+            if (s_isInt16 && reader.TryGetInt32(out int int16) && int16 >= short.MinValue && int16 <= short.MaxValue)
+            {
+                value = (TValue)Enum.ToObject(valueType, int16);
+                return true;
+            }
+
+            if (s_isByte && reader.TryGetUInt32(out uint ubyte8) && ubyte8 >= byte.MinValue && ubyte8 <= byte.MaxValue)
```

Comment:
Also, I would imagine this to be the next most common (so maybe position this after the int32 check). The rest of the ordering can probably remain as is (and we can re-order in the future if we get more data points to justify it).

Reply:
done